### PR TITLE
Support Arrays

### PR DESCRIPTION
**Pull request by emlyn**
Will need some tests, but thought I'd open the PR already to start getting feedback. Does this look reasonable?

### CODE REVIEW

File: pythonosc/osc_message_builder.py [hunk 1]
```diff
@@ -46,6 +49,16 @@ def args(self):
     """Returns the (type, value) arguments list of this message."""
     return self._args
 
+  def valid_type(self, arg_type):
```
Comment (attwad) [inline pythonosc/osc_message_builder.py]: _valid_type, this doesn't need to be called outside of this class

File: pythonosc/osc_message_builder.py [hunk 2]
```diff
@@ -56,25 +69,44 @@ def add_arg(self, arg_value, arg_type=None):
     Raises:
       - ValueError: if the type is not supported.
     """
-    if arg_type and arg_type not in self._SUPPORTED_ARG_TYPES:
+    if arg_type and not self.valid_type(arg_type):
       raise ValueError(
           'arg_type must be one of {}'.format(self._SUPPORTED_ARG_TYPES))
```
Comment (attwad) [inline pythonosc/osc_message_builder.py]: this error message also need to be fixed now that you're checking for another error case

File: pythonosc/osc_message_builder.py [hunk 3]
```diff
@@ -56,25 +69,44 @@ def add_arg(self, arg_value, arg_type=None):
...
+      arg_type = self.get_arg_type(arg_value)
+    if isinstance(arg_type, list):
+      self._args.append((self.ARG_TYPE_ARRAY_START, None))
+      for v, t in zip(arg_value, arg_type):
+        self.add_arg(v, t)
+      self._args.append((self.ARG_TYPE_ARRAY_STOP, None))
+    else:
+      self._args.append((arg_type, arg_value))
+
+  def get_arg_type(self, arg_value):
```
Comment (attwad) [inline pythonosc/osc_message_builder.py]: ditto prefix with "_" and name accordingly: _guess_arg_type

File: pythonosc/osc_message.py [hunk 4]
```diff
@@ -49,11 +52,25 @@ def _parse_datagram(self):
           val = True
         elif param == "F": # False.
           val = False
+        elif param == "[": # Array start.
+          a = []
```
Comment (attwad) [inline pythonosc/osc_message.py]: please avoid single letter vars, at least 3 like "arg"/"val" is the minimum acceptable ;)

File: pythonosc/osc_message.py [hunk 5]
```diff
@@ -33,8 +33,11 @@ def _parse_datagram(self):
       if type_tag.startswith(','):
         type_tag = type_tag[1:]
 
+      params = []
+      param_stack = [params]
       # Parse each parameter given its type.
       for param in type_tag:
+        have_val = True
```
Comment (attwad) [inline pythonosc/osc_message.py]: I'd prefer an array specific approach here instead of this generic one:
is_within_array = False
or something like that

**Comment by emlyn** [comment]
Thanks for the feedback, I've addressed your comments and added some tests.

**Comment by attwad** [comment]
Thanks a lot!

**Pull request merged**